Add visualization of open controlled gates for mpl #3957

enavarro51 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Fix #3957

### Details and comments

For open controlled gates, the text drawer shows which control qubits are open or closed. This fix adds the same function for the mpl drawer. In addition the params for the gates are now shown on the gates in the same manner as they would be for a cu3 gate.

Code also had to be added to compute the proper layer_width. The default sections of this code resulted in excessively large layer widths. The new code sets layer_width to 1 if there are 10 or fewer chars in the displayed params, and to 2 if greater than 10. The max is 15 chars. As an example,

```
from qiskit import QuantumRegister, QuantumCircuit
from qiskit.extensions import U3Gate, U2Gate
import numpy as np

pi = np.pi
qr = QuantumRegister(4, 'q')
circuit = QuantumCircuit(qr)
circuit.append(U2Gate(pi/8,pi/8).control(3, ctrl_state=2),
               [qr[0], qr[3], qr[2], qr[1]])
circuit.append(U3Gate(3*pi/4,3*pi/4,3*pi/4).control(3, ctrl_state='101'),
               [qr[0], qr[3], qr[2], qr[1]])
circuit.draw(output='mpl')
```

![image](https://user-images.githubusercontent.com/16268251/79054573-f1fab000-7bfa-11ea-82ff-22d18ee5dfdf.png)

enavarro51 commented:
@1ucian0, @ajavadia After submitting this, I noticed #4052 and realized the mpl drawer has this problem as well. I have a fix for it, but in the process a question came up. There seems to be an inconsistency in matplotlib.py related to face colors for gates. Should a ControlledGate have the color of the base gate, as in a C3XGate would have the color of an X gate, or should all ControlledGates be the same color? And if they are different colors, should the line colors and the control qubits use those colors as well?

enavarro51 commented:
And along these lines, in the text drawer, the cx and ccx gates display the target cubit as an ordinary X gate. In latex and mpl, it is displayed as a circle with a plus in it. So given the following code, which is the equivalent of a ccx should this use the X gate or the circle-plus?
`circuit.append(XGate().control(2, ctrl_state=3),
               [qr[0], qr[1], qr[2]])`
And if the answer is the circle-plus, then I assume that if ctrl_state != 3, then we would use the X gate since the open control bits would not make this a Toffoli. Thanks.

enavarro51 commented:
Oh, the joys of digging deeply into code. The controlled-Z is handled as a special case and instead of displaying the target qubit as a Z-gate, it displays the target qubit just like a closed control qubit. Any reason for this? Obviously makes it a little hard to distinguish control from target.
`circuit.append(ZGate().control(3, ctrl_state='101'),
               [qr[0], qr[3], qr[2], qr[1]])`
q_0: ─■─
      │ 
q_1: ─■─
      │ 
q_2: ─■─
      │ 
q_3: ─o─

ajavadia commented:
> There seems to be an inconsistency in matplotlib.py related to face colors for gates.

@enavarro51 for gate colors please see this comment: https://github.com/Qiskit/qiskit-terra/pull/4038#issuecomment-605658480
This PR recently fixed all colors, so I think it should be ok, unless you see something wrong.

> So given the following code, which is the equivalent of a ccx should this use the X gate or the circle-plus?
circuit.append(XGate().control(2, ctrl_state=3), [qr[0], qr[1], qr[2]])
And if the answer is the circle-plus, then I assume that if ctrl_state != 3, then we would use the X gate since the open control bits would not make this a Toffoli.

I think it should use circle-plus regardless of control state. Circle-plus is just a common circuit notation for X. The reason this is not done in the text drawer is that it doesn't render nicely.

> The controlled-Z is handled as a special case and instead of displaying the target qubit as a Z-gate, it displays the target qubit just like a closed control qubit. Any reason for this? Obviously makes it a little hard to distinguish control from target

This is correct. This gate is symmetric, it does not alter the gate if you change control and target. So this is a common way of displaying this fact.

enavarro51 commented:
Ok, got the colors. Also I'll leave the controlled-Z as is, and all Contolled-X will use the circle plus for the target. Thanks again.

ajavadia commented:
@enavarro51 it looks like cswap gates have inconsistent colors. if you could fix that it would be great. It should be all purple (because it is non-Clifford).

<img width="129" alt="image" src="https://user-images.githubusercontent.com/8622381/79082275-8136bf00-7cf2-11ea-8573-8e622a7c7c98.png">

enavarro51 commented:
Got it. Let me know if there are any other anomalies since I'm doing a complete roll through the code.

ajavadia commented:
I also see CU1 and RZZ are displayed in black. These should be purple. Only measurement and reset should be black. 

<img width="56" alt="image" src="https://user-images.githubusercontent.com/8622381/79088151-b998c600-7d0f-11ea-9f89-d46e79891737.png">

<img width="67" alt="image" src="https://user-images.githubusercontent.com/8622381/79088160-c3bac480-7d0f-11ea-82bd-6b8d585be0fd.png">

enavarro51 commented:
Ok, got those. Is there a reason those 2 gates are done with sidetext instead of in a box like almost all the other gates? Just trying to go for as much consistency as I can.
Also wanted to know if an open cx, cy, etc. is a valid gate, such as,
`circuit.append(XGate().control(1, ctrl_state='0'), [qr[0], qr[1]])`
Thanks.

enavarro51 commented:
Further comments. Updated all the colors to match #4038 list, except for U1, U2, & U3 which were in red and were left that way. rzz and cu1 were the only 2 gates with sidetext, so changed these to the standard - rzz looks like rxx, ryy, etc. and cu1 looks like cu3. Corrected colors on the controlled-swap gate. Fixed some errors in matching the display to the ctrl_state.
Tested all gates listed in qiskit/extensions/standard.

ajavadia commented:
Thanks. 
U1 u2 u3 we can keep the way it is because those are special IBM basis gates (at least for now). 
For cu1 and rzz, the intention of drawing them like that was that they are symmetric gates, and that’s how they usually render inquantum circuit diagrams. So I think they should be left the way they were. 

> On Apr 13, 2020, at 1:35 PM, Edwin Navarro <notifications@github.com> wrote:
> 
> Further comments. Updated all the colors to match #4038 list, except for U1, U2, & U3 which were in red and were left that way. rzz and cu1 were the only 2 gates with sidetext, so changed these to the standard - rzz looks like rxx, ryy, etc. and cu1 looks like cu3. Corrected colors on the controlled-swap gate. Fixed some errors in matching the display to the ctrl_state.
> Tested all gates listed in qiskit/extensions/standard.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

enavarro51 commented:
Ok. I'll put those 2 back the way they were and recommit.

ajavadia commented on qiskit/visualization/qcstyle.py:
```diff
@@ -96,8 +96,8 @@ def __init__(self):
             'meas': non_gate_color
         }
         self.latexmode = False
-        self.fold = None  # To be removed after 0.10 is released
```
according to this comment, `fold` should just be removed from `qcstyle`. it is now a top-level arg to `circuit.draw()`.

enavarro51 replied:
Ok. Here are 3 outputs. Do you want me to remove all references to fold in qcstyle?

![image](https://user-images.githubusercontent.com/16268251/79283699-748aa600-7e6d-11ea-8ac1-1aadaaa3026f.png)
![image](https://user-images.githubusercontent.com/16268251/79283783-b9aed800-7e6d-11ea-8722-b06616e0e837.png)
![image](https://user-images.githubusercontent.com/16268251/79283810-d0552f00-7e6d-11ea-86fc-e29d3913f0ee.png)

ajavadia commented:
Yes I think fold should be removed from style. 

The colors look but the U3 wodth seems to cause overflows. Can this be fixed?

> On Apr 14, 2020, at 7:37 PM, Edwin Navarro <notifications@github.com> wrote:
> 
> @enavarro51 commented on this pull request.
> 
> In qiskit/visualization/qcstyle.py:
> 
> > @@ -96,8 +96,8 @@ def __init__(self):
>              'meas': non_gate_color
>          }
>          self.latexmode = False
> -        self.fold = None  # To be removed after 0.10 is released
> Ok. Here are 3 outputs. Do you want me to remove all references to fold in qcstyle?
> 
> 
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

enavarro51 commented:
Will take care of fold. The overflow in the U3 and others sometimes is due to autoscaling. This is not something I changed. Here's a sample output without autoscaling,
![image](https://user-images.githubusercontent.com/16268251/79288393-778c9300-7e7b-11ea-829a-f3fb41b72c6a.png)
I assumed this was expected behavior. Apparently the text is not scaled when the rest of the image is. I can look into this if you'd like. I noticed that the fold for mpl looks pretty clunky compared to the fold for text_drawer. Let me know if you want me to address this or move it to another issue.

ajavadia commented:
got it. let's open another issue for that then.

I just noticed the control for CX is green (not blue like the rest of the gate). The other controls, e.g. CY or CZ look fine. Can you fix this and then I can approve

<img width="78" alt="image" src="https://user-images.githubusercontent.com/8622381/79291303-4e302f00-7e9c-11ea-9259-13310a627d36.png">

<img width="99" alt="image" src="https://user-images.githubusercontent.com/8622381/79291322-57210080-7e9c-11ea-8dcf-75d01450d343.png">

enavarro51 commented:
Ok. I'll take care of these and recommit in the morning.
